sync/profile.go

```diff
@@ -1,45 +1,440 @@
+}
+
+// ProfileImportResult is returned by successful import operations.
+type ProfileImportResult struct {
+	ImportResult `json:",inline"`
+
+	ReplacesProfiles []string `json:"replacesProfiles"`
+}
+
+func registerProfileAPI() error {
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Name:        "Export App Profile",
+		Description: "Exports app fingerprints, settings and metadata in a share-able format.",
+		Path:        "sync/profile/export",
+		Read:        api.PermitAdmin,
+		Write:       api.PermitAdmin,
+		Parameters: []api.Parameter{{
+			Method:      http.MethodGet,
+			Field:       "id",
+			Description: "Specify scoped profile ID to export.",
+		}},
+		BelongsTo: module,
+		DataFunc:  handleExportProfile,
+	}); err != nil {
+		return err
+	}
+
+	if err := api.RegisterEndpoint(api.Endpoint{
+		Name:        "Import App Profile",
+		Description: "Imports full app profiles, including fingerprints, setting and metadata from the share-able format.",
+		Path:        "sync/profile/import",
+		Read:        api.PermitAdmin,
+		Write:       api.PermitAdmin,
+		Parameters: []api.Parameter{
+			{
+				Method:      http.MethodPost,
+				Field:       "allowReplace",
+				Description: "Allow replacing existing profiles.",
+			}, {
+				Method:      http.MethodPost,
+				Field:       "validate",
+				Description: "Validate only.",
+			}, {
+				Method:      http.MethodPost,
+				Field:       "reset",
+				Description: "Replace all existing settings.",
+			}, {
+				Method:      http.MethodPost,
+				Field:       "allowUnknown",
+				Description: "Allow importing of unknown values.",
+			}},
+		BelongsTo:  module,
+		StructFunc: handleImportProfile,
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func handleExportProfile(ar *api.Request) (data []byte, err error) {
+	var request *ProfileExportRequest
+
+	// Get parameters.
+	q := ar.URL.Query()
+	if len(q) > 0 {
+		request = &ProfileExportRequest{
+			ID: q.Get("id"),
+		}
+	} else {
+		request = &ProfileExportRequest{}
+		if err := json.Unmarshal(ar.InputData, request); err != nil {
+			return nil, fmt.Errorf("%w: failed to parse export request: %w", ErrExportFailed, err)
+		}
+	}
+
+	// Check parameters.
+	if request.ID == "" {
+		return nil, errors.New("missing parameters")
+	}
+
+	// Export.
+	export, err := ExportProfile(request.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return serializeExport(export, ar)
+}
+
+func handleImportProfile(ar *api.Request) (any, error) {
+	var request *ProfileImportRequest
+
+	// Get parameters.
+	q := ar.URL.Query()
+	if len(q) > 0 {
+		request = &ProfileImportRequest{
+			ImportRequest: ImportRequest{
+				ValidateOnly: q.Has("validate"),
+				RawExport:    string(ar.InputData),
+				RawMime:      ar.Header.Get("Content-Type"),
+			},
+			AllowUnknown: q.Has("allowUnknown"),
+			AllowReplace: q.Has("allowReplace"),
+		}
+	} else {
+		request = &ProfileImportRequest{}
+		if err := json.Unmarshal(ar.InputData, request); err != nil {
```

Comment:
Allocating an empty struct for `request` is not required, json.Unmarshal will do it anyway

Reply:
> Unmarshal parses the JSON-encoded data and stores the result in the value pointed to by v. If v is nil or not a pointer, Unmarshal returns an InvalidUnmarshalError.